src/string_object.cpp

```diff
@@ -2517,4 +2517,32 @@ Value StringObject::sum(Env *env, Value val) {
 
     return Value::integer(sum);
 }
+
+Value StringObject::try_convert(Env *env, Value val) {
+    auto to_str = "to_str"_s;
+
+    if (val->is_string()) {
+        return val;
+    }
+
+    if (!val->respond_to(env, to_str)) {
+        return NilObject::the();
+    }
+
+    auto conversion = val->send(env, to_str);
+
+    if (conversion->is_string() || conversion->is_nil()) {
+        return conversion;
+    }
+
+    auto original_item_class_name = val->klass()->inspect_str();
+    auto new_item_class_name = conversion->klass()->inspect_str();
+    env->raise(
+        "TypeError",
+        "can't convert {} to Array ({}#to_str gives {})",
+        original_item_class_name,
+        original_item_class_name,
+        new_item_class_name);
+}
```

Comment:
This whole method is very similar to `Object::to_str`, I wonder if we can just call that method inside here?

Reply:
I didn't previously think to do that, but just tried it and it failed one of the specs:
```
Failed specs:
String.try_convert
  sends #to_str to the argument and returns the result if it's nil
    can't convert MockObject to String (MockObject#to_str gives NilClass) (TypeError)
      spec/core/string/try_convert_spec.rb:22:in `try_convert'
```
It seems that the behavior is subtly different in how nils are handled between the different approaches.

Reply:
Ahhh, ok! Very well, let's just use your code then.